### PR TITLE
fix the problem that cpu_fallback for aten::triu_indices on custom device crashed

### LINKED ISSUE #121289: cpu_fallback for aten::triu_indices on custom device crashed
### 🐛 Describe the bug

My team is working on adding a custom backend for pytorch and found that falling back to CPU process of aten::triu_indices did not correctly execute. Following is the reproduce code:

```
import torch 
import torch_mlu  
torch.triu_indices(3, 3, device='mlu')
```
Error Message
```
/projs/framework/chentianyi1/pytorch/catch/test/[tri.py:3](http://tri.py:3/): UserWarning: The operator 'aten::triu_indices' is not currently supported on the MLU backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /projs/framework/chentianyi1/pytorch/catch/torch_mlu/csrc/aten/MLUFallback.cpp:49.)
  print(torch.triu_indices(3,3,device='mlu'))
Traceback (most recent call last):
  File "/projs/framework/chentianyi1/pytorch/catch/test/[tri.py](http://tri.py/)", line 3, in <module>
    print(torch.triu_indices(3,3,device='mlu'))
RuntimeError: device_or_default(device_opt).type() == DeviceType::CPU INTERNAL ASSERT FAILED at "/projs/framework/chentianyi1/pytorch/aten/src/ATen/EmptyTensor.cpp":256, please report a bug to PyTorch.
```
This probably because in cpu_fallback (https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUFallback.cpp#L80), optional device arguement was not converted to CPU but passed to cpu kernel directly. And triggered errror in empty_cpu(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/EmptyTensor.cpp#L256).

### Versions

Collecting environment information...
PyTorch version: 2.1.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.27.7
Libc version: glibc-2.31

Python version: 3.10.8 (main, Dec 12 2023, 17:13:35) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-172-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 57 bits virtual
CPU(s):                             112
On-line CPU(s) list:                0-111
Thread(s) per core:                 2
Core(s) per socket:                 28
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              106
Model name:                         Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz
Stepping:                           6
Frequency boost:                    enabled
CPU MHz:                            2599.992
CPU max MHz:                        3100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4000.00
Virtualization:                     VT-x
L1d cache:                          2.6 MiB
L1i cache:                          1.8 MiB
L2 cache:                           70 MiB
L3 cache:                           84 MiB
NUMA node0 CPU(s):                  0-27,56-83
NUMA node1 CPU(s):                  28-55,84-111
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] optree==0.10.0
[pip3] torch==2.3.0a0+git36dccc2
[pip3] torch-mlu==1.18.0+torch2.1.0
[conda] Could not collect

cc @bdhirsh

### PR DESCRIPTION
**Pull request by chentianyi16**
Fixes #121289

### CODE REVIEW

File: aten/src/ATen/native/CPUFallback.cpp [hunk 1]
```diff
@@ -183,9 +187,13 @@ void cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack, bool
   const auto& num_returns = schema_returns.size();
   auto returns = torch::jit::last(stack, num_returns);
   const auto returns_begin = stack->size() - num_returns;
-
-  c10::optional<c10::Device> tgt_device =
-      compute_target_device(tensor_args, tensorlist_args);
+  
+  c10::optional<c10::Device> tgt_device = c10::nullopt;
+  if (compute_target_device(tensor_args, tensorlist_args) != c10::nullopt) {
```
Comment (ezyang) [inline aten/src/ATen/native/CPUFallback.cpp:192]: nit: assign this to a variable so you don't recompute it?

**Comment by ezyang** [comment]
Needs a test

File: aten/src/ATen/native/CPUFallback.cpp (hunk 1 above)
Comment (chentianyi16) [reply aten/src/ATen/native/CPUFallback.cpp]: fixed

**Comment by chentianyi16** [comment]
> Needs a test

A test is added in the latest commit

**Comment by ezyang** [comment]
I think this implementation is not quite right. In particular, I think the Device argument detection should in fact take precedence over device detection on inputs; for example, if I somehow have foobar(cpu_tensor, device='cuda'), it seems pretty clear that I expect the output device to be cuda. So you should have tgt_device be defined way up top, if you see a device take that as the tgt device, and then if there's nothing there, then you try to infer it from tensor inputs.

**Comment by chentianyi16** [comment]
> I think this implementation is not quite right. In particular, I think the Device argument detection should in fact take precedence over device detection on inputs; for example, if I somehow have foobar(cpu_tensor, device='cuda'), it seems pretty clear that I expect the output device to be cuda. So you should have tgt_device be defined way up top, if you see a device take that as the tgt device, and then if there's nothing there, then you try to infer it from tensor inputs.

Thanks for your reply and I've modified my PR based on your comment.

**Review by ezyang — APPROVED** [review]
thanks, very slick now

**Comment by ezyang** [comment]
lint

**Comment by chentianyi16** [comment]
> lint

fixed and all checks passed.

**Pull request closed without merging**